ui-core/src/components/inputs/datePicker/useDatePicker.ts

```diff
@@ -0,0 +1,73 @@
+import { useState, useRef, useEffect } from 'react';
+import useClickOutside from '../../hooks/useOutsideClick';
+import { useModalPosition } from '../../hooks/useModalPosition';
+import { computeNewSelectedSlot, formatInputValue, formatValueToSlot } from './utils';
+import { DatePickerProps, RangeDatePickerProps, SingleDatePickerProps } from './DatePicker';
+
+const MODAL_HORIZONTAL_OFFSET = -24;
+const MODAL_VERTICAL_OFFSET = 3;
+
+export default function useDatePicker({
+  inputProps,
+  value,
+  isRangeMode,
+  onDateChange,
+}: DatePickerProps) {
+  const [showPicker, setShowPicker] = useState(false);
+  const [inputValue, setInputValue] = useState(formatInputValue(value, isRangeMode));
+  const [selectedSlot, setSelectedSlot] = useState(formatValueToSlot(value, isRangeMode));
+
+  const calendarPickerRef = useRef<HTMLDivElement>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+  useClickOutside(calendarPickerRef, (e) => {
+    // Do not close the picker if any children in input wrapper is clicked.
+    // This wrapper include, the input itself, the trailing content (which contains the calendar icon) and the leading content
+    if (inputRef.current && inputRef.current.parentElement?.contains(e.target as Node)) return;
+    setShowPicker(false);
+  });
+  const { calculatePosition, modalPosition } = useModalPosition(
+    inputRef,
+    calendarPickerRef,
+    MODAL_VERTICAL_OFFSET,
+    MODAL_HORIZONTAL_OFFSET
+  );
+
+  const handleInputClick = (e: React.MouseEvent<HTMLInputElement, MouseEvent>) => {
+    setShowPicker(true);
+    inputProps.onClick?.(e);
+  };
+
+  const handleDayClick = (clickedDate: Date) => {
+    if (isRangeMode) {
+      const newSelectedSlot = computeNewSelectedSlot(
+        clickedDate,
+        value as RangeDatePickerProps['value']
+      );
+      (onDateChange as RangeDatePickerProps['onDateChange'])(clickedDate, newSelectedSlot);
+    } else {
+      (onDateChange as SingleDatePickerProps['onDateChange'])(clickedDate);
```

Comment:
Why did you add cast here ? Thanks to the typing, it's not need normally
```suggestion
      const newSelectedSlot = computeNewSelectedSlot(
        clickedDate,
        value
      );
      onDateChange(clickedDate, newSelectedSlot);
    } else {
      onDateChange(clickedDate);
```